File: leetcode/SingleNumber.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var singleNumber = function (nums) {
	let flag = {};
	let rs = 0;
	for (let i = 0; i < nums.length; i++) {
		let num = nums[i];
		if (flag[num] !== 1) flag[num] = 1;
		else flag[num] = -1;
		rs += flag[num] * num;
	}
	return rs;
};
singleNumber([2,2,1])

var singleNumber = function (nums) {
	const dict = {}
	nums.forEach(v => {
		dict[v] = dict[v] + 1 || 1
	})
	for(let num in dict) {
		if(dict[num] === 1) return num
	}
}

var singleNumber = function (nums) {
	let rs = nums[0]
	for(let i = 1;i< nums.length;i++) {
		rs = rs ^ nums[i]
	}
	return rs
}
/**
 * Best practice using XOR
 * class Solution(object):
 *  def singleNumber(self, nums):
 *      a = 0
 *      for i in nums:
 *          a ^= i
 *      return a
 */

File: leetcode/singleNumber.py
class Solution:
    def singleNumber(self, nums: List[int]) -> int:
        return single_number_set(nums)

# Brute force find each num O(n^2)

# Set O(n)


def single_number_set(nums):
    s = set()

    for n in nums:
        if n not in s:
            s.add(n)
        else:
            s.remove(n)
    return s.pop()

# A XOR A = 0 and the XOR operator is commutative
# 0 XOR A = A
# 0 ^ res ^ a ^ b ^ a ^ b = res


def single_number_xor(nums):
    res = 0
    for n in nums:
        res = res ^ n
    return res
